### ISSUE TITLE
[ENH] Architectural Proposal: Data Pipeline Subclassing for Foundation Models (TTM Case Study)

### ISSUE THREAD

**Issue by StrikerEureka34** [jdb78/pytorch-forecasting#2184]
**Related:** [#1959](https://github.com/sktime/pytorch-forecasting/issues/1959) [#2051](https://github.com/sktime/pytorch-forecasting/issues/2051)

### Is your feature request related to a problem? Please describe.

I've been looking into what it would take to wire IBM's [TinyTimeMixer (TTM)](https://github.com/ibm-granite/granite-tsfm/blob/main/tsfm_public/models/tinytimemixer/modeling_tinytimemixer.py) into the v2 pipeline. As part of that, I did a fairly thorough sweep, of `_tslib_data_module.py` and `_base_model_v2.py` against TTM's actual `forward()` signature. Here's what I found.
&ensp;

#### Architecture Sweep Findings

**Finding 1: Exogenous features work differently in TTM, and target may get double-counted**

This is the big one. TTM has **no `exogenous_values` parameter** in its [forward()](https://github.com/ibm-granite/granite-tsfm/blob/main/tsfm_public/models/tinytimemixer/modeling_tinytimemixer.py). Instead, exogenous features go into `past_values` as additional channels. The config uses [prediction_channel_indices](https://github.com/ibm-granite/granite-tsfm/blob/main/tsfm_public/models/tinytimemixer/modeling_tinytimemixer.py) and [exogenous_channel_indices](https://github.com/ibm-granite/granite-tsfm/blob/main/tsfm_public/models/tinytimemixer/modeling_tinytimemixer.py) to tell the model which channels are which.

This maps directly to the channel-ordering requirements from #2051 [(thread here)](https://github.com/sktime/pytorch-forecasting/issues/2051#issuecomment-4053508328) @agobbifbk's DummyDataset example stacks `[target, known_future_covs, past_only_covs]` into a single tensor with strict ordering:
1. Targets first
2. Past-only covariates next
3. Known-future covariates last and their order must match the order in `future_values`

The `_TTMDataset.__getitem__` I'm proposing enforces exactly this ordering.

Two sub-problems:
- **(a) Mask asymmetry:** [cont_known_mask](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L169-L173) is applied to future_cont to filter known features only, but history_cont at [L148](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L148) keeps **all** features. TTM needs the same channel count on both sides.
- **(b) Target deduplication:** If a user puts the target column in both y and x features, the target shows up twice after channel-concat, once as history_target, once inside history_cont. The exogenous mask needs to explicitly exclude target columns.

&nbsp;

**Finding 2: Observed mask doesn't handle NaNs and is per-timestep only**

The `history_mask` from [_preprocess_data](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L570) is just `timestep <= cutoff_time`. But TTM's `past_observed_mask` is (batch, seq_len, num_input_channels) **per-channel**. 
It expects 0 where values are missing/NaN. Without this, NaNs propagate through attention and produce NaN loss.

&nbsp;

**Finding 3: collate_fn hardcodes Tslib-specific keys**

The [collate_fn](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L845-L862) references keys like history_cont, history_cat, etc. A TTM dataset returning `past_values`, `future_values`, `past_observed_mask` would hit KeyError.

&nbsp;

**Finding 4: future_values and loss routing need careful handling**

[BaseModel.training_step](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model_v2.py#L178) unpacks `x, y = batch`. HF's TTM computes its own internal loss when `return_loss=True`, so we need `return_loss=False` to prevent double loss.
The target portion goes in y for PTF's loss; `future_values` can be omitted from x entirely since TTM doesn't use it when `return_loss=False`.
&nbsp;

**Finding 5: No caching on _preprocess_data**

[_preprocess_data](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L540-L614) does pandas -> numpy -> tensor conversion and is called on every single item retrieval. With a stride-1 sliding window on a 1000-step series, the same series gets preprocessed ~900 times per epoch. 
This is especially relevant given the memory/compute concerns @agobbifbk raised in #2051 and was validated in one of my benchmark tests (D2-style precomputation on a 10k-length, 20-feature COSS series ≈ 8 GB CPU RAM , [here](https://github.com/sktime/pytorch-forecasting/issues/2051#issuecomment-3959533549).

The lazy getitem approach I am proposing actually helps here, caching the tensor per-series avoids both the repeated pandas -> tensor overhead *and* the memory explosion of fully precomputed windows.

&nbsp;
**Finding 6: Scaling is a stub; TTM has its own internal scaler**

The `_target_normalizer` (RobustScaler) is instantiated in [__init__](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L337-L340) but never called. TTM uses its own scalers (TinyTimeMixerMeanScaler/TinyTimeMixerStdScaler). For initial integration, we can let TTM handle scaling.

&nbsp;

**Finding 7: Channel layout is fine** 
TTM expects channels-last (batch, seq_len, channels), which already matches `_TslibDataset` output. No permutation needed.
&nbsp;
**Finding 8: kwargs dead code** 
[TslibDataModule.__init__](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L313) accepts kwarg arguments and stores them but never uses them.
&nbsp;
**Finding 9: Random series split** 
[setup()](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L700) uses `torch.randperm` for train/val/test splits. Pre-existing, already has a TODO in source.

&ensp;

### Describe the solution you'd like

All of these can be addressed with **one small backwards-compatible edit** to the existing code, plus a new file. No changes to `_base_model_v2.py`, no changes to existing tests.

This follows the same reuse-over-fork philosophy in which the existing TslibDataModule handles ~80% of what foundation models need (`_preprocess_data`, `_create_windows`, `setup()`, series splitting). 
Rather than duplicating that logic or adding model-specific branches, we subclass it.
This requirement was a must have and was explicitly discussed in the recent weekly meetups by the maintainers.

&nbsp;
**The one edit** introduce a `_dataset_class` class attribute on TslibDataModule:

[setup()](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/data/_tslib_data_module.py#L727-L760) currently hardcodes `_TslibDataset` in four places. The change is **1 addition + 4 substitutions**:

```python
class TslibDataModule(L.LightningDataModule):
    _dataset_class = _TslibDataset  # NEW: default preserves existing behavior

    def setup(self, stage=None):
        ...
        # Replace 4 occurrences of _TslibDataset(...) with:
        self.train_dataset = self._dataset_class(dataset=..., data_module=self, ...)
```

&nbsp;

**Everything else is a new file** (`_fm_data_module.py`) with `TTMDataModule(TslibDataModule)` that sets `_dataset_class = _TTMDataset` and provides:
- A `_TTMDataset` that overrides `__getitem__` to do the channel-concat (following @agobbifbk's ordering: targets -> past-only -> known-future), per-channel NaN masking, and target deduplication
- A generic `collate_fn` that doesn't hardcode key names
- A `_preprocess_data` wrapper with per-series caching (avoids both repeated conversion *and* memory explosion from full precomputation)
- A typed config dataclass instead of the dead `**kwargs`

The parent's `_preprocess_data`, `_create_windows`, `setup()`, and series splitting are all inherited as-is. 
This same pattern might work for future FMs (Chronos, Moirai, TimesMixer, etc.)  just a new subclass with its own `__getitem__` and `collate_fn` if otherwise a new requirement specific to a FM is discovered.

&ensp;

### Describe alternatives you've considered
- **A completely separate `FMDataModule`** not inheriting from `TslibDataModule` over 80% of the pipeline (`_preprocess_data`, `_create_windows`, `setup()`, splitting) is reusable. Duplicating it means two copies to maintain. This was a concern raised in #2051 as well.

- **A `_preprocess_for_model` hook on `BaseModel`** pushes data-formatting into the model layer. Tensor layout is a data concern, and this makes it harder to inspect raw dataloader output for debugging.

&ensp;

### Additional context

I've traced every tensor shape through the full pipeline (preprocess -> `__getitem__` -> collate ->  `training_step` -> TTM forward -> predict) and confirmed dimensionality is consistent.
Edge cases like univariate, multi-target, no exogenous features, short series, and context length mismatches all check out.

One nuance: during real inference (not evaluation), `future_values` won't be available. The `_TTMDataset` handles this via a stage flag, omitting it from `x` during predict.

&nbsp;
I've been researching this heavily over the last two weeks and trying to align it with the directions discussed in the recent meetups. I've mapped out the actual code-level solutions and class hierarchies, which I'm happy to drop in follow-up comments.

Looking forward to hear what you think of this direction! @agobbifbk @PranavBhatP @phoeenniixx

**Comment by StrikerEureka34** [comment]
One quick note regarding inference with this setup:

I logged a small bug last week [(#2156)](https://github.com/sktime/pytorch-forecasting/issues/2156) where the v2 PredictCallback currently misreads TslibDataModule batches. We'll probably need to apply that small fix so that the proposed TTMDataModule can run `.predict()` without crashing. 

It's a super quick, contained fix, I’d be more than happy to open a PR for #2156 in case needed.
